### PR TITLE
gdb: add support for debuginfod

### PR DESCRIPTION
**Pull request by baloo**
###### Motivation for this change

Adds support for gdb to grab debuginfo off a debuginfod

see #112454

gdb will issue requests like:
```
GET /buildid/d62d19ec82fe63de07f367941445bd9f658a5312/debuginfo HTTP/1.1
Host: localhost:1234
Accept: */*
Accept-Encoding: deflate, gzip
User-Agent: elfutils/0.180,Linux/x86_64,nixos/21.03.git.dfd114708a5
```
To the url specified in DEBUGINFOD_URLS environment variable.

We can then defer that to nix-locate and nix-store like:
```
$ nix-locate d6/2d19ec82fe63de07f367941445bd9f658a5312.debug
glibc.debug                                      79,104 x /nix/store/a91mg1lzassyjqk6sffl8181ci9bjg5z-glibc-2.32-35-debug/lib/debug/.build-id/d6/2d19ec82fe63de07f367941445bd9f658a5312.debug
```

That would download on-demand the debug symbols.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by baloo** [comment]
(the efitools commits are submited in a separated PR, but this requires the changes made there)

File: pkgs/development/tools/misc/gdb/default.nix [hunk 1]
```diff
@@ -48,7 +49,8 @@ stdenv.mkDerivation rec {
 
   buildInputs = [ ncurses readline gmp mpfr expat libipt zlib guile ]
     ++ lib.optional pythonSupport python3
-    ++ lib.optional doCheck dejagnu;
+    ++ lib.optional doCheck dejagnu
+    ++ lib.optional debuginfodSupport elfutils;
```
Comment (baloo) [inline pkgs/development/tools/misc/gdb/default.nix:53]: that does not make much good if elftools does not provide libdebuginfod here.

Should I do something like:
```
let elfutils = elfutils.override{enableDebuginfod = debuginfodSupport; };
```
 or something?

**Comment by baloo** [comment]
I'm not too sure why that generates a derivation all that different to be honest. Debuginfod is set to false, I did not expect to trigger a system-wide rebuild.

Should I go to change base branch to staging-next then?

**Comment by Mic92** [comment]
This normal for gdb. I think it rebuilds at least rustc.

**Comment by baloo** [comment]
?!

**Comment by Mic92** [comment]
Done.

**Comment by baloo** [comment]
to note, the debuginfodSupport is default to false here. The tests does not really tests my changes here.

File: pkgs/development/tools/misc/elfutils/default.nix [hunk 2]
```diff
@@ -52,17 +54,25 @@ stdenv.mkDerivation rec {
     })
   ] ++ lib.optional stdenv.hostPlatform.isMusl [ ./musl-error_h.patch ];
 
+  outputs = [ "bin" "dev" "out" "man" ];
```
Comment (Mic92) [inline pkgs/development/tools/misc/elfutils/default.nix:57]: This change might trigger the rebuilds.

**Review by nicknovitski — APPROVED** [review]

**Pull request merged**